src/code/z_sub_s.c

```diff
@@ -214,17 +213,260 @@ void SubS_UpdateFlags(u16* flags, u16 setBits, u16 unsetBits) {
+        weights[j] = weightsTemp[order - 1][j];
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_sub_s/func_8013B6B0.s")
+/**
+ * Computes the X and Z component of the position to move to in time based paths
+ *
+ * @param[out] x
+ * @param[out] z
+ * @param[in] progress see `SubS_TimePathing_ComputeProgress`
+ * @param[in] order the order of the interpolation i.e. the number of points in the interpolation, max is 10
+ * @param[in] waypoint the current waypoint
+ * @param[in] points the path's points
+ * @param[in] weightArray see `SubS_TimePathing_FillWeightArray`
+ */
+void SubS_TimePathing_ComputeTargetPosXZ(f32* x, f32* z, f32 progress, s32 order, s32 waypoint, Vec3s points[],
+                                         f32 weightArray[]) {
+    f32 xPos;
+    f32 zPos;
+    f32 weights[11];
+    f32 weightedX;
+    f32 weightedZ;
+    f32 weightedTotal;
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_sub_s/func_8013B878.s")
+    SubS_TimePathing_ComputeWeights(order, progress, waypoint, weightArray, weights);
+    weightedTotal = 0.0f;
+    weightedZ = 0.0f;
+    weightedX = 0.0f;
+
+    for (i = 0; i < order; i++) {
+        xPos = points[waypoint - order + i + 1].x;
+        zPos = points[waypoint - order + i + 1].z;
+
+        weightedX += weights[i] * xPos;
+        weightedZ += weights[i] * zPos;
+        weightedTotal += weights[i];
+    }
+    *x = weightedX / weightedTotal;
+    *z = weightedZ / weightedTotal;
+}
+
+/**
+ * Updates a time based path that an actor follows by:
+ *  - Computing the X and Z components of the next point to move to
+ *  - Updating the waypoint
+ *  - Updating the time
+ *
+ * @param[in] path
+ * @param[out] progress see `SubS_TimePathing_ComputeProgress`
+ * @param[in,out] elapsedTime how much time has passed
+ * @param[in] waypointTime how much time per each waypoint
+ * @param[in] totalTime how much time the path should take to travel
+ * @param[in,out] waypoint the current waypoint, this and the previous two points will be used to compute the targetPos
+ * @param[in] weightArray see `SubS_TimePathing_FillWeightArray`
+ * @param[out] targetPos the computed position to move to
+ * @param[in] timeSpeed how fast time moves
+ *
+ * @return s32 returns true when the end has been reached.
+ */
+s32 SubS_TimePathing_Update(Path* path, f32* progress, s32* elapsedTime, s32 waypointTime, s32 totalTime, s32* waypoint,
+                            f32 weightArray[], Vec3f* targetPos, s32 timeSpeed) {
+    Vec3s* points = Lib_SegmentedToVirtual(path->points);
+    s32 state;
+    f32 endX;
+    f32 endZ;
+    s32 reachedEnd = false;
+
+    if (*waypoint >= path->count) {
+        state = 2;
+    } else {
+        state = SubS_TimePathing_ComputeProgress(progress, *elapsedTime, waypointTime, totalTime, path->count,
+                                                 SUBS_TIME_PATHING_ORDER, weightArray);
+    }
+
+    switch (state) {
+        case 1: // Haven't reached the end of the path
```

Comment:
Don't really think its worth it for an internal function, but if you think it is better I can add one.

Reply:
My impression was that it was used several times in here. And if you're commenting on it every time, it might as well be an enum.

Reply:
Its only used once, though it gets documented twice since it is both the return value from `SubS_TimePathing_ComputeProgress` and used in this switch.

Reply:
Right, and since it's used in 2 functions I think it's worth it.